[IMPROVEMENT] Change bottom sheet lib

IlarionHalushka opened the pull request:
@RocketChat/ReactNative
Closes #759

This PR aims to replace Action Sheet with Reanimated Bottom Sheet on message actions.

As you could notice I am not a great designer so suggestions are welcome :)
big screen Iphone X: 
![image](https://user-images.githubusercontent.com/25078524/56762091-ac559500-67a7-11e9-9213-2d4e88f5fcad.png)

small screen Iphone SE: 
![image](https://user-images.githubusercontent.com/25078524/56762083-a790e100-67a7-11e9-9ebc-8d80a8cd556e.png)

diegolmello commented:
@IlarionHalushka Cool!!
We're not trying to copy Slack, but here's how they do it:
<img src="https://user-images.githubusercontent.com/804994/56762309-f2920080-6775-11e9-8002-49ea86899554.jpeg" width="350" />

The idea is to show an amount of items and let the user swipe up if he wants to show the rest.
Also add the icons on the left.

For small screens: the tallest person in the world can have the smallest device, so he still needs a way to tap on the buttons :)
Leave the buttons in the same size and don't worry about it for now.

diegolmello commented:
@IlarionHalushka this is how our design looks like for commands, autocomplete, etc:
<img src="https://user-images.githubusercontent.com/804994/56762640-adba9980-6776-11e9-9aab-8977c2c4878d.png" width="350" />

Use your lib with our design and you're fine :)

IlarionHalushka commented:
I updated the styles and added close on click-out. 

When Longpress performed on message:
![image](https://user-images.githubusercontent.com/25078524/56804606-02215000-682f-11e9-9717-a6616df7caad.png)

When user scrolled the bottom action sheet from bottom to top:
![image](https://user-images.githubusercontent.com/25078524/56804629-1ebd8800-682f-11e9-9e43-7e3641582ac6.png)

diegolmello commented:
@IlarionHalushka Looking so much better already!
Fix android build and I'll do a review.

IlarionHalushka commented:
Improved the transitions: https://drive.google.com/file/d/1V4JmmbyzgIqmTaLjBfdjbLdzMbvPF8Q1/view?usp=sharing

diegolmello commented on app/containers/MessageActions.js:
```diff
@@ -264,79 +276,72 @@ export default class MessageActions extends React.Component {
 	render() {
 		return (
-			null
+			<BottomSheet
```
A backdrop is missing.
You can use something like this: https://github.com/RocketChat/Rocket.Chat.ReactNative/blob/develop/app/views/RoomsListView/SortDropdown.js#L121

IlarionHalushka replied:
unfortunately the bottomsheet doesn't have function `onPositionChange` or something similar, but I did manage to do backdrop using `Animated.Code`

diegolmello commented on app/views/RoomView/index.js:
```diff
@@ -131,6 +130,8 @@ export default class RoomView extends LoggedView {
 				EventEmitter.addEventListener('connected', this.handleConnected);
 			}
+
+			navigation.addListener('willBlur', () => this.hideActionSheet());
```
I think this is not needed as RoomView will be removed from the stack, but test it.

If it's necessary, check if it's opened and move it to `componentWillUnmount`.
This is similar to `editCancel` and `replyCancel`.

IlarionHalushka replied:
unfortunately the bottomsheet doesn't have function `onPositionChange` or something similar, but I did manage to do backdrop using `Animated.Code`

IlarionHalushka replied:
It's necessary, because when I open Bottom Action Sheet in RoomView and then go to Threads Single View (thread chat) then ActionSheet is still displayed there and same happens when changing RoomViews(chats)

IlarionHalushka replied:
And also I tried to move it to `componentWillUnmount` but as we use StackNavigator - the view in not unmounted when opening Threads View or Settings view and bottom sheet stays constantly opened.

diegolmello commented on ios/RocketChatRN.xcodeproj/project.pbxproj:
```diff
@@ -1684,20 +1693,32 @@
 				CLANG_CXX_LIBRARY = "libc++";
 				CLANG_ENABLE_MODULES = YES;
 				CLANG_ENABLE_OBJC_ARC = YES;
+				CLANG_WARN_BLOCK_CAPTURE_AUTORELEASING = YES;
```
I think you all these `CLANG*` are from your Xcode setup and shouldn't be commited.

IlarionHalushka replied:
removed in latest commit

diegolmello commented on app/views/RoomView/index.js:
```diff
@@ -490,15 +499,15 @@ export default class RoomView extends LoggedView {
-				{room._id && showActions
-					? <MessageActions room={room} user={user} />
+				{room._id
+					? <MessageActions room={room} user={user} setRef={ref => this.setRef(ref)} />
```
Creating arrow functions on render is a bad practice and it should be avoided.
This creates functions on every render.

You should always be using the class function directly like this:
`setRef={this.setRef}`

IlarionHalushka replied:
did bind `this.setRef` and then used like you suggested `setRef={this.setRef}`

diegolmello commented on ios/RocketChatRN.xcodeproj/project.pbxproj:
```diff
@@ -22,6 +22,7 @@
 		146834051AC3E58100842450 /* libReact.a in Frameworks */ = {isa = PBXBuildFile; fileRef = 146834041AC3E56700842450 /* libReact.a */; };
 		24A2AEF2383D44B586D31C01 /* libz.tbd in Frameworks */ = {isa = PBXBuildFile; fileRef = 06BB44DD4855498082A744AD /* libz.tbd */; };
 		2C800DF680F8451599E80AF1 /* libSafariViewManager.a in Frameworks */ = {isa = PBXBuildFile; fileRef = 1D3BB00B9ABF44EA9BD71318 /* libSafariViewManager.a */; };
+		3EAC15CC2271955100AD43B2 /* libRNReanimated.a in Frameworks */ = {isa = PBXBuildFile; fileRef = 3EAC15CB2271937800AD43B2 /* libRNReanimated.a */; };
```
Am I wrong or did you tried to add reanimated from imports and from pods?
If I'm right, it's just a matter of changing `Podfile` and lines like this one is trash.

IlarionHalushka replied:
I did link the library manually using Xcode, but not sure what all these lines are about... :(

diegolmello commented on android/app/build.gradle:
```diff
@@ -243,4 +244,4 @@ task copyDownloadableDepsToLibs(type: Copy) {
 apply plugin: 'com.google.gms.google-services'
-com.google.gms.googleservices.GoogleServicesPlugin.config.disableVersionCheck = true
\ No newline at end of file
+com.google.gms.googleservices.GoogleServicesPlugin.config.disableVersionCheck = true
```
Unnecessary change

IlarionHalushka replied:
this was made by Webstorm, returned back

IlarionHalushka commented on app/containers/MessageActions.js:
```diff
@@ -264,79 +276,72 @@ export default class MessageActions extends React.Component {
-		const { actionsHide } = this.props;
-		actionsHide();
+	hideActionSheet() {
+		this.bottomSheetRef.current.snapTo(hiddenBottomSheet);
```
fixed

IlarionHalushka commented on app/containers/MessageActions.js:
```diff
@@ -138,36 +161,25 @@ export default class MessageActions extends React.Component {
 			return null;
 		}
-	}
+	};
```
removed

pranavpandey1998official commented:
@IlarionHalushka  can you resolve conflicts

pranavpandey1998official commented:
@IlarionHalushka the vibration are setting off immediately after opening the room and not after we long press ..

pranavpandey1998official commented:
In my testing no button work

pranavpandey1998official commented:
where did the separator go , It was nice as user was able to distinguish between boundaries of button 
![WhatsApp Image 2019-05-11 at 6 27 00 PM](https://user-images.githubusercontent.com/44601530/57569979-7c022d80-741a-11e9-850c-3d587690ac49.jpeg)

pranavpandey1998official commented:
its showing start all the time (unstar is not showing up)

pranavpandey1998official commented:
pin is not working

IlarionHalushka commented:
**1.** `covers full screen screen in landscape` . fixed 
![image](https://user-images.githubusercontent.com/25078524/57571404-67239b00-7416-11e9-887d-c5ad5d2fa431.png)

**2.** `where did the separator go , It was nice as user was able to distinguish between boundaries of button` - fixed.

**3.** `its showing start all the time (unstar is not showing up)` - currently I don't have any good idea how to change the titles because the bottomSheet is rendered once when the RoomView is opened. 

**4.** `pin is not working` - is not connected to BottomSheet, action is called but 400 Not Authorized response is returned from Back-end (same behaviour on current testflight app). [Issue](https://github.com/RocketChat/Rocket.Chat/issues/13454)
